silicon/backends/rabbitmq.hh

```diff
@@ -0,0 +1,260 @@
+		serialize(response_type * r, T const & res) const
+		{
+		}
+	};
+
+	struct context
+	{
+		template <typename... O>
+		context(unsigned short port, O &&... opts)
+		{
+			auto options = D(opts...);
+			auto hostname = options.get(s::_hostname, std::string("localhost"));
+			auto username = options.get(s::_username, std::string("guest"));
+			auto password = options.get(s::_password, std::string("guest"));
+
+			conn = amqp_new_connection();
+			socket = amqp_tcp_socket_new(conn);
+
+			if (!socket)
+				throw std::runtime_error("creating TCP socket");
+
+			status = amqp_socket_open(socket, hostname.c_str(), port);
+			if (status)
+				throw std::runtime_error("opening TCP socket");
+
+			die_on_amqp_error(amqp_login(conn, "/", 0, 131072, 0, AMQP_SASL_METHOD_PLAIN, username.c_str(), password.c_str()), "Loggin in");
+			amqp_channel_open(conn, 1);
+			die_on_amqp_error(amqp_get_rpc_reply(conn), "Opening channel");
+		}
+
+		int status;
+		amqp_socket_t * socket = nullptr;
+		amqp_connection_state_t conn;
+		std::vector<amqp_bytes_t> queuenames;
+	};
+
+	template <typename A, typename M, typename... O>
+	auto
+	make_context(A const & api, M const & mf, unsigned short port, O &&... opts)
+	{
+		auto ctx = context(port, opts...);
+		auto options = D(opts...);
+		auto exchange = options.get(s::_exchange, std::string(""));
+
+		foreach(api) | [&] (auto& m)
+		{
+			iod::static_if<is_tuple<decltype(m.content)>::value>(
+					[&] (auto m) { // If sio, recursion.
+						throw std::runtime_error("FIXME: m.content is a tuple, not handle today");
+					},
+					[&] (auto m) { // Else, register the procedure.
+						std::stringstream bindingkey;
+
+						foreach(m.route.path) | [&] (auto e)
+						{
+
+							iod::static_if<is_symbol<decltype(e)>::value>(
+									[&] (auto e2) {
+										bindingkey << std::string("/") + e2.name();
+									},
+									[&] (auto e2) {
+										// FIXME: dynamic symbol
+									},
+							e);
+						};
+
+						amqp_queue_declare_ok_t * r = amqp_queue_declare(ctx.conn, 1, amqp_empty_bytes, 0, 0, 0, 1, amqp_empty_table);
+						amqp_bytes_t queuename;
+
+						die_on_amqp_error(amqp_get_rpc_reply(ctx.conn), "Declaring queue");
+						queuename = amqp_bytes_malloc_dup(r->queue);
+						if (queuename.bytes == NULL)
+						{
+							throw std::runtime_error("Out of memory while copying queue name");
+						}
+
+						amqp_queue_bind(ctx.conn, 1, queuename, amqp_cstring_bytes(exchange.c_str()), amqp_cstring_bytes(bindingkey.str().c_str()), amqp_empty_table);
+						die_on_amqp_error(amqp_get_rpc_reply(ctx.conn), "Binding queue");
+
+						amqp_basic_consume(ctx.conn, 1, queuename, amqp_empty_bytes, 0, 1, 0, amqp_empty_table);
+						die_on_amqp_error(amqp_get_rpc_reply(ctx.conn), "Consuming");
+
+						ctx.queuenames.emplace_back(queuename);
+					},
+			m);
+		};
+
+		return ctx;
+	}
+
+	template <typename A, typename M, typename... O>
+	auto
+	serve(A const & api, M const & mf, unsigned short port, O &&... opts)
```

Comment:
If this implements the RMQ consumer, it would be nice that the function name reflecting it, maybe consume instead, like in the python API.